crates/sp-domains/src/offline_operators.rs

```diff
@@ -0,0 +1,188 @@
+//     τ = 0.1% => ln(1000)  ≈ 6.907755278982137  -> inner ≈ 6_907_755_278_982_137_000
+// - To decide throughput relevance before calling this (cheap filter), compute
+//   μ_floor = floor(S * p) and require μ_floor >= E_relevance (e.g., >= 10 for τ=1%).
+
+#[cfg(test)]
+mod tests;
+
+use crate::StakeWeight;
+use crate::bundle_producer_election::calculate_threshold;
+use core::cmp::max;
+use frame_support::pallet_prelude::{Decode, Encode, TypeInfo};
+use frame_support::{Deserialize, Serialize};
+use sp_arithmetic::traits::{One, SaturatedConversion, Saturating, UniqueSaturatedInto, Zero};
+use sp_arithmetic::{FixedPointNumber, FixedU128};
+use sp_core::U256;
+
+/// For τ = 1%: ln(1/τ) = ln(100) ≈ 4.605170185988092
+/// Represent in FixedU128 by multiplying by 1e18 and rounding.
+pub const LN_1_OVER_TAU_1_PERCENT: FixedU128 = FixedU128::from_inner(4_605_170_185_988_092_000);
+
+/// Threshold for number of bundles operator can produce based on their stake share.
+/// Operators whose expected bundles from their threshold < E_BASE will not be checked
+/// since they are not throughput relevant.
+pub const E_BASE: u64 = 3;
+
+// Exact per-slot win probability is p = threshold / 2^128.
+// Convert that to FixedU128 (1e18 scaling) using U256 intermediates.
+fn p_from_threshold(threshold: u128) -> FixedU128 {
+    // inner = floor(threshold * 1e18 / 2^128)
+    let num = U256::from(threshold).saturating_mul(U256::from(FixedU128::DIV));
+    let den = U256::from(1u128) << 128; // 2^128
+    FixedU128::from_inner((num / den).unique_saturated_into())
+}
+
+/// Chernoff lower-tail threshold r for Binomial(S, p), with false-positive at most τ.
+///
+/// Inputs:
+/// - slots_in_epoch: S (number of slots in the epoch)
+/// - p_slot: per-slot success probability p in FixedU128 (1e18 scaling)
+/// - ln_one_over_tau: ln(1/τ) in FixedU128 (1e18 scaling), precomputed off-chain
+///
+/// Returns:
+/// - r in [0, S], as u64
+///
+/// Formula:
+///   μ = S * p
+///   r = floor( μ - sqrt( 2 * μ * ln(1/τ) ) ), clamped to [0, S]
+///
+/// Guarantees:
+///   P_honest(X < r) <= τ for X ~ Binomial(S, p). Conservative (safe) bound.
+///
+/// Usage:
+///   let r = chernoff_threshold_fp(S, p, ln_1_over_tau);
+///   if observed_bundles < r { exclude } else { keep }
+fn chernoff_threshold_fp(
+    slots_in_epoch: u64,
+    p_slot: FixedU128,
+    ln_one_over_tau: FixedU128,
+) -> Option<u64> {
+    if slots_in_epoch == 0 || p_slot.is_zero() {
+        return Some(0);
+    }
+
+    if p_slot >= FixedU128::one() {
+        return Some(slots_in_epoch);
+    }
+
+    // μ = S * p
+    let mu = p_slot.saturating_mul(FixedU128::saturating_from_integer(slots_in_epoch));
+
+    // t = sqrt( 2 * μ * ln(1/τ) )
+    let t = FixedU128::from(2u128)
+        .saturating_mul(mu)
+        .saturating_mul(ln_one_over_tau)
+        .try_sqrt()?;
+
+    // r = floor( μ - t ), clamped to [0, S]
+    Some(
+        if mu > t {
+            mu.saturating_sub(t).into_inner() / FixedU128::DIV
```

Comment:
Personally, I would write this as `floor()` to make the code much clearer, even if I had to write a floor trait myself. But it's up to you.

(This PR contains 3 manual floor() and 1 manual ceil() calculations.)